src/native/libs/System.Native/pal_networking.c

```diff
@@ -1660,6 +1660,11 @@ int32_t SystemNative_Connect(intptr_t socket, uint8_t* socketAddress, int32_t so
     return err == 0 ? Error_SUCCESS : SystemNative_ConvertErrorPlatformToPal(errno);
 }
 
+#if defined(__linux__) && !defined(TCP_FASTOPEN_CONNECT)
+// fixup if compiled against old Kernel headers.
+// Can be removed once we have at least 4.11
+#define TCP_FASTOPEN_CONNECT 30
```

Comment:
We're relying on getsockopt below failing when this is defined but not understood?

Reply:
yes. If the call fails e.g. not supported by given Kernel we would fall-back to traditional connect and data would be sent after handshake.